### PR TITLE
feat: initial codegen

### PR DESCRIPTION
**Pull request by tjjfvi**
Makes progress towards #226 

This cleans up the codegen code somewhat, though it's still not the prettiest. It also adds codegen for codecs, as well as initial stub codegen for pallets and their storage and calls.

This has only been minimally manually tested; automatic testing should be added, but we need to figure out the best approach for this.

### CODE REVIEW

**Comment by harrysolovay** [comment]
Tested with Polkadot and the output looks correct.

This is a great foundation: types, corresponding codecs (each explicitly typed to `$Codec<T>` and named in accordance with their idx, specialized codecs drawn into scope, tsdocs wherever possible, and even the beginnings of extrinsic / storage bindings! Awesome work @tjjfvi.

Would love to be able to use it through something like...

```sh
deno run -A -r https://deno.land/x/capi/codegen.ts \
  --src="wss://rpc.polkadot.io" \
  --out="polkadot.ts"
```

File: codegen/genMetadata.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                : "[]",
+            ],
+            ["value", getRawCodecPath(tys[entry.value]!)],
+          ),
+        ],
+      });
+    }
+    if (pallet.calls) {
+      const ty = tys[pallet.calls.ty]! as M.Ty & M.UnionTyDef;
+      const isStringUnion = ty.members.every((x) => !x.fields.length);
```
Comment (harrysolovay) [inline codegen/genMetadata.ts:70]: A pallet, all the callables of which accept no arguments? I suppose it's possible ;)

File: codegen/genMetadata.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      for (const call of ty.members) {
+        const typeName = isStringUnion ? S.string(call.name) : getPath(tys, ty)! + "." + call.name;
+        const [params, data]: [S, S] = call.fields.length
+          ? call.fields[0]!.name
+            ? [`value: Omit<${typeName}, "type">`, ["{ ...value, type:", S.string(call.name), "}"]]
```
Comment (harrysolovay) [inline codegen/genMetadata.ts:75]: Very nice to say the least

File: codegen/typeVisitor.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    taggedUnion(ty) {
+      const path = getPath(tys, ty)!;
+      const name = getName(path);
+      decls.push({
+        path,
+        code: [
+          makeDocComment(ty.docs),
+          ["export type", name, "="],
+          ty.members.map(({ fields, name: type, docs }) => {
+            let props: [S, S, S][];
```
Comment (harrysolovay) [inline codegen/typeVisitor.ts:47]:
```suggestion
            let props: [comment: S, name: S, type: S][];
```

Comment (harrysolovay) [reply codegen/typeVisitor.ts:47]: ^ labels will make this a bit more legible to future contributors

File: codegen/utils.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
+import * as M from "../frame_metadata/mod.ts";
+
+export const importSource = new URL("../mod.ts", import.meta.url).toString();
```
Comment (harrysolovay) [inline codegen/utils.ts:3]: How do we want to deal with Capi import specifier across project types?

- a new string param in the CLI?
- attempt to infer based on current process (does `Deno` global exist, misc.)
- ...

File: codegen/utils.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  return _getName(ty);
+
+  function _getName(ty: M.Ty): string | null {
+    if (ty.type === "Primitive") {
+      return ty.kind;
+    }
+    if (ty.type === "Compact") {
+      return null;
+    }
+    if (ty.path.at(-1) === "Era") return "Era";
```
Comment (harrysolovay) [inline codegen/utils.ts:42]: Might a pallet developer expose a different type with the name `Era`?

File: codegen/utils.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+export function getName(path: string) {
+  return path.split(".").at(-1)!;
+}
+
+export function makeDocComment(docs: string[]) {
+  docs = docs.map((x) => x.replace(/^\s*\n\s*|\s*\n\s*$/, "").replace(/\s*\n\s*/g, " "));
+  if (!docs.length) return "";
+  if (docs.length === 1) return `/** ${docs[0]!.trim()} */\n`;
```
Comment (harrysolovay) [inline codegen/utils.ts:63]: AWESOME

File: codegen/utils.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    "$" + parts.at(-1)![0]!.toLowerCase() + parts.at(-1)!.slice(1),
+  ].join(".");
+}
+
+export function printDecls(decls: Decl[]) {
+  const namespaces: Record<string, Decl[]> = {};
+  const done: Decl[] = [];
+  for (const { path, code } of decls) {
+    if (path.includes(".")) {
+      const ns = path.split(".")[0]!;
```
Comment (harrysolovay) [inline codegen/utils.ts:89]: Nit. Worth reusing `path.split(".")`?

File: codegen/utils.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  done.sort((a, b) =>
+    a.path.startsWith("_") !== b.path.startsWith("_")
+      ? a.path.startsWith("_") ? -1 : 1
+      : a.path < b.path
+      ? -1
+      : a.path > b.path
+      ? 1
+      : 0
+  );
+  return [...new Set(done.map((x) => S.toString(x.code)))].join("\n");
```
Comment (harrysolovay) [inline codegen/utils.ts:116]: Why is it necessary to turn this into a set? Assuming the metadata is valid, are duplicates possible?

File: deno.jsonc [hunk 9]
```diff
@@ -38,6 +38,6 @@
     "test:update": "deno task test -- -- --update",
     "mdbook:watch": "mdbook watch -o",
     "bench": "deno bench -A --no-check=remote --unstable",
-    "typegen": "deno task run typegen/typegen.ts"
+    "codegen": "deno task run codegen/codegen.ts"
```
Comment (harrysolovay) [inline deno.jsonc:41]: Eventually `deno run -A ./main.ts`?

**Review by harrysolovay — CHANGES_REQUESTED** [review]
Looking very nice. Review is mostly nits––pretty much merge-ready.

File: codegen/codecVisitor.ts [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+      return addCodecDecl(ty, ["$.map(", this.visit(key), ",", this.visit(val), ")"]);
+    },
+    set(ty, val) {
+      return addCodecDecl(ty, ["$.set(", this.visit(val), ")"]);
+    },
+    era(ty) {
+      return addCodecDecl(ty, "$era");
+    },
+    circular(ty) {
+      return ["$.deferred(() =>", getName(getRawCodecPath(ty)), ")"];
```
[the remark this answers is not in the document]
Comment (tjjfvi) [reply codegen/codecVisitor.ts:168]: That was on the `DeriveCodec` side, and only for the case of unions – this usage of deferred is mandatory.

File: codegen/utils.ts (hunk 8 above)
Comment (tjjfvi) [reply codegen/utils.ts:116]: It is, though I don't recall why

File: codegen/utils.ts (hunk 4 above)
Comment (tjjfvi) [reply codegen/utils.ts:3]: #258

File: codegen/utils.ts (hunk 5 above)
Comment (tjjfvi) [reply codegen/utils.ts:42]: This is also an issue with the codec derivation. #259

**Review by harrysolovay — APPROVED** [review]
Awesome work @tjjfvi

**Pull request merged**